crates/primitives/src/result.rs

```diff
@@ -1,26 +1,39 @@
-use crate::{Address, Bytes, Log, State, U256};
+use crate::{Address, Bytes, ChainSpec, Log, State, U256};
 use core::fmt;
 use std::{boxed::Box, string::String, vec::Vec};
 
 /// Result of EVM execution.
-pub type EVMResult<DBError> = EVMResultGeneric<ResultAndState, DBError>;
+pub type EVMResult<ChainSpecT, DBError> = EVMResultGeneric<ResultAndState<ChainSpecT>, DBError>;
 
 /// Generic result of EVM execution. Used to represent error and generic output.
 pub type EVMResultGeneric<T, DBError> = core::result::Result<T, EVMError<DBError>>;
 
 #[derive(Debug, Clone, PartialEq, Eq)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-pub struct ResultAndState {
+pub struct ResultAndState<ChainSpecT: ChainSpec> {
```

Comment:
The cast functions were a remnant from another attempt to achieve this. I've removed them in https://github.com/bluealloy/revm/pull/1378/commits/26dec59ba2c7af2207045088cbcf9bf3a5304d24.

W.r.t using the generic `ChainSpec` or `HaltReason`; up to you. I chose `HaltReason` in case some hardforks introduce custom `SuccessReason`s down the line.